CorsixTH/Lua/objects/doors/swing_door_right.lua

```diff
@@ -38,24 +38,55 @@ local SwingDoor = _G["SwingDoor"]
 function SwingDoor:SwingDoor(hospital, object_type, x, y, direction, etc)
   self.is_master = object_type == object
   self:Door(hospital, object_type, x, y, direction, etc)
-  if self.is_master then
-    local slave_type = "swing_door_left"
-    self.slave = self.world:getObject(x - 1, y, slave_type) or self.world:getObject(x, y - 1, slave_type) or nil
-    self.slave:setAsSlave(self)
-    assert(self.slave.master, "Swing doors are not paired at " .. x .. "," .. y)
-  end
+  self.paired = false
+  self:pairDoors(x, y)
   self.old_anim = self.th:getAnimation()
   self.old_flags = self.th:getFlag()
   local --[[persistable:swing_door_creation]] function _() end -- Stub from Oct 2023, savegame version 181
 end
 
---[[ Makes the door mimic its master when it comes to hover cursor and what happens
-when the player clicks on it.
+--! Links the master to the slave swing door. This allows the slave to mimic the
+--! master in interactions
+--!param x (num) Co-ordinate
+--!param y (num) Co-ordinate
+function SwingDoor:pairDoors(x, y)
+  -- Because we don't know the build order of each door we must accomodate for this
+  if self.is_master then -- The master will check for an adjacent slave
+    local slave_type = "swing_door_left"
+    self.slave = self.world:getObject(x - 1, y, slave_type) or
+        self.world:getObject(x, y - 1, slave_type)
+    if self.slave then
+      self.slave.master = self
+      self.paired = true
+      self.slave.master.paired = true
+    end
+  else -- The slave will check for an adjacent master
+    local master_type = "swing_door_right"
+    self.master = self.world:getObject(x + 1, y, master_type) or
+        self.world:getObject(x, y + 1, master_type)
+    if self.master then
+      self.master.slave = self
+      self.paired = true
+      self.master.slave.paired = true
+    end
+  end
+  -- The second run of this function must check there was a successful pairing
+  self:checkPaired(x, y)
+end
 
-!param swing_door (Door) The master door to mimic.
-]]
-function SwingDoor:setAsSlave(swing_door)
-  self.master = swing_door
+--! Checks to see if the swing doors have paired correctly
+function SwingDoor:checkPaired(x, y)
+  if self.paired then
+    assert((self.master and self.master.slave == self) or
+        (self.slave and self.slave.master == self),
+        "Swing doors did not pair at (" .. x .. ", " .. y .. ")")
+  else -- The other door should not yet exist
```

Comment:
This is reached when pairDoors is run on a door without its pair? I can't imagine how, but it's worth catching. 👍

Reply:
Yeah... I feel like because of the way this function works we actually have to test for all cases:
- Case A: The doors report themsleves and paired and we should make sure that the correct pairing happened
- Case B: The doors report not paired, and we should make sure that the adjacent door doesn't exist.